### PR TITLE
Finishing the TMState Invariants

### PR DESCRIPTION
**Pull request by Grendel-Grendel-Grendel**
This PR should add all of the necessary invariant testing. This checks the length of the `Notebook` and `FocusList` and also compares each widget in each to make sure they are the same. It does this by turning each into al ist of widgets and iterating `widgetEq` through each. If it encounters a mismatch, it throws an error. 

I made sure there were no performance problems using the list method by opening up an unreasonable number of tabs and seeing if the program suffered any problems running its checks, and it looks like there was no problem.

### CODE REVIEW

**Comment by cdepillabout** [comment]
So overall I think this looks good.  There are a few small style changes I would make, but other than that, it is good.

Do you want me to do a code review and leave comments for all the changes I would make?  Or would you rather I just merge it in and fix everything myself?

**Comment by Grendel-Grendel-Grendel** [comment]
The former would be wonderful. Thanks.

File: src/Termonad/Types.hs [hunk 1]
```diff
@@ -306,6 +327,42 @@ invariantTMState' tmState =
...
+    invariantTabsAllMatch :: IO (Maybe TMStateInvariantErr)
+    invariantTabsAllMatch = do
+      --Turns a FocusList and Notebook into two lists of widgets and compares each widget for equality
+      let tmNote = tmNotebook $ tmStateNotebook tmState
+          focusList = tmNotebookTabs $ tmStateNotebook tmState
+          flList = fmap tmNotebookTabTermContainer $ toListFL focusList
+      noteList <- notebookToList tmNote
+      tabsMatch noteList flList 0
+      where
+        tabsMatch :: (IsWidget a, IsWidget b) => [a] -> [b] -> Int -> IO (Maybe TMStateInvariantErr)
```
[the remark this answers is not in the document]
Comment (Grendel-Grendel-Grendel) [reply src/Termonad/Types.hs]: What is the "this" here? `tabsMatch`? Maybe you should just stick it in there as i find it exceedingly difficult to read. Everything else you've commented should be fixed now.

[the remark this answers is not in the document]
Comment (LSLeary) [reply src/Termonad/Types.hs]: @Grendel-Grendel-Grendel This is all written unaided and I haven't looked much at the surrounding code, so there may be errors, but a more basic cleanup might look like this:

```haskell
tabsMatch :: (IsWidget a, IsWidget b) => [a] -> [b] -> Int -> IO (Maybe TMStateInvariantErr)
tabsMatch xs ys = go (zip xs ys)
  where
    go ((x, y) : xys) i = widgetEq x y >>= \case
      True  -> tabsMatch xys (i + 1)
      False -> pure . Just $ TabsDoNotMatch (TabAtIndexDifferent i)
    go _ _ = pure Nothing
```

Another version which recognises that we're folding and tracking index state manually, then eliminates that work to arrive at essentially what cdepillabout wrote:

```haskell
tabsMatch :: (IsWidget a, IsWidget b) => [a] -> [b] -> IO (Maybe TMStateInvariantErr)
tabsMatch xs ys = foldr go (pure Nothing) (zip3 xs ys [0..])
  where
    go (x, y, i) acc = widgetEq x y >>= \case
      True  -> acc
      False -> pure . Just $ TabsDoNotMatch (TabAtIndexDifferent i)
```

[the remark this answers is not in the document]
Comment (LSLeary) [reply src/Termonad/Types.hs]: @Grendel-Grendel-Grendel, a couple of extra points.

Using `asum` over `Maybe a` values that have been determined according to a monadic predicate is just a roundabout form of `findM`. If we had as standard in `base` (and I believe we should):

```haskell
findM :: Monad m => (a -> m Bool) -> [a] -> m (Maybe a)
findM f (x:xs) = f x >>= \case
  True  -> pure (Just x)
  False -> findM f xs
findM _ _ = pure Nothing

(<&&>) :: (Functor f, Functor g) => f (g a) -> (a -> b) -> f (g b)
(<&&>) = flip $ (<$>) . (<$>)
```

Then I would write:

```haskell
tabsMatch :: (IsWidget a, IsWidget b) => [a] -> [b] -> IO (Maybe TMStateInvariantErr)
tabsMatch xs ys = findM (\(x, y, _) -> not <$> widgetEq x y) (zip3 xs ys [1..])
  <&&> \(_, _, i) -> TabsDoNotMatch (TabAtIndexDifferent i)
```

You can get a similar looking version using `filterM` if you don't mind the fact that unlike `findM` it must traverse and accumulate monadic effects from the whole list. But imo the `asum` option is better than that.

Also, as of 087b2dd there are still trailing spaces on lines 210, 297 and 349. You should set up your editor to highlight them.

**Comment by cdepillabout** [comment]
@Grendel-Grendel-Grendel Thanks for fixing up a lot of the stuff I mentioned!  Sorry for my lack of explanation with `foldr`, but LSLeary gave a very comprehensive explanation.  Once you fix up the code like he suggests, I will go ahead and merge this PR in!

**Comment by Grendel-Grendel-Grendel** [comment]
It's fixed up now. I didn't use the solution with foldM, however, but rather the penultimate solution.

**Comment by cdepillabout** [comment]
Thanks for cleaning this up!

**Pull request merged**